_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,140 @@
+
+Unfortunately at 2am, we got frantic calls from customers saying that our App is broken! Looking at the logs, you find lots of `NoSuchMethodError` is being thrown by some code in `A`! 
+
+![Binary incompatibility after upgrading]({{ site.baseurl }}/resources/images/library-author-guide/after_update.png){: style="width: 280px; margin: auto; display: block;"}
+
+Why did we get a `NoSuchMethodError`? Remember that `A v1.0.0` is compiled with `C v1.0.0` and thus calls methods available in `Cv1.0.0`. While `B` and
+our App has been recompiled with available classes/methods in `C v2.0.0`, `A v1.0.0`'s bytecode hasn't changed - it still calls the same method that is now missing in `C v2.0.0`!
+
+This situation can only be resolved by ensuring that the chosen version of `C` is binary compatible with all other evicted versions of `C`. In this case, we need a new version of `A` that depends
+on `C v2.0.0` (or any other future `C` version that is binary compatible with `C v2.0.0`).
+
+Now imagine if our App is more complex with lots of dependencies themselves depending on `C` (either directly or transitively) - it becomes extremely difficult to upgrade any dependencies because it now
+pulls in a version of `C` that is incompatible with the rest of the versions of `C` in our dependency tree! In the example below, we cannot upgrade `D` because it will transitively pull in `C v2.0.0`, causing breakages
+due to binary incompatibility. This inability to upgrade any packages without breaking anything is common known as **Dependency Hell**.
+
+![Dependency Hell]({{ site.baseurl }}/resources/images/library-author-guide/dependency_hell.png)
+
+How can we, as library authors, spare our users of runtime errors and dependency hell?
+
+* Use **Migration Manager** (MiMa) to catch unintended binary compatibility breakages before releasing a new library version
+* **Avoid breaking binary compatibility** through careful design and evolution of your library interfaces
+* Communicate binary compatibility breakages clearly through **versioning**
+
+## MiMa - Check Binary Compatibility with Previous Library Versions 
+
+The [Migration Manager for Scala](https://github.com/typesafehub/migration-manager) (MiMa) is a tool for diagnosing binary incompatibilities between different library versions.
+
+When run standalone in the command line, it will compare the .class files in the two provided JARs and report any binary incompatibilities found. Most library authors use the [SBT plugin](https://github.com/typesafehub/migration-manager/wiki/Sbt-plugin)
+to help spot binary incompatibility between library releases. (Follow the link for instructions on how to use it in your project)
+
+## Designing for Evolution - without breaking binary compatibility
+
+TODO
+
+## Versioning Scheme - Communicate binary compatibility breakages
```

Comment:
What if you try to explain what a versioning scheme is and maybe quote the semantic versioning spec? I feel like a small introduction here would make the reading much much nicer for beginners.

Reply:
I didn't want to mention semantic versioning at all because it brings in the wrong correlation and might confuse the reader. 

`MAJOR version when you make incompatible API changes,` 

In the versioning scheme we're recommending, major version bump has nothing to do with API changes. (From my point of view, API changes = source incompatibility)

Reply:
but yeah, we can briefly explain what versioning schemes are

Reply:
> I didn't want to mention semantic versioning at all because it brings in the wrong correlation and might confuse the reader.
  
I'm not sure about this. The versioning scheme that we recommend **is** a more constrained subset of semantic versioning. I feel mentioning semantic versioning here is a must, it's a widely used standard. We should make it obvious that our versioning scheme enforces more information in every Scala version, which is extremely advantageous.

> (From my point of view, API changes = source incompatibility)
  
Not really, API changes happen because of both binary and source incompatible changes.